terraform/openvdc/provider.go

```diff
@@ -0,0 +1,71 @@
+package openvdc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+	"os/exec"
+	"strings"
+)
+
+func Provider() terraform.ResourceProvider {
+	return &schema.Provider{
+		Schema: map[string]*schema.Schema{},
+
+		ResourcesMap: map[string]*schema.Resource{
+			"openvdc_instance": OpenVdcInstance(),
+		},
+	}
+
+}
+
+func RunCmd(name string, arg ...string) (*bytes.Buffer, *bytes.Buffer, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	cmd := exec.Command(name, arg...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+
+	return &stdout, &stderr, err
+}
+
+func ParseJson(cmdStdOut *bytes.Buffer) (string, error) {
+	var parsed map[string]interface{}
+
+	err := json.Unmarshal(cmdStdOut.Bytes(), &parsed)
+	if err != nil {
+		fmt.Println("Uh-oh! You had an error!")
+		return "", err
+	}
+	var x map[string]interface{}
+	x = parsed["instance"].(map[string]interface{})
+
+	x = x["last_state"].(map[string]interface{})
+
+	return x["state"].(string), nil
+
+}
+
+func CheckInstanceTerminated(cmdStdOut *bytes.Buffer) (bool, error) {
+	status, err := ParseJson(cmdStdOut)
+
+	if err != nil {
+		return false, err
+	}
+
+	if strings.Compare(status, "FAILED") == 0 {
+		return false, fmt.Errorf("Instance is in FAILED state!")
```

Comment:
Unless I'm missing something this in combination with [line 100 of resouce_openvdc_instance.go](https://github.com/axsh/openvdc/pull/218/files#diff-654da0062ce0dead0ec37848cdcc86beR100) will still cause `terraform destroy` to fail if the instance is already in FAILED state.

Reply:
Right. I had thought that this was wanted. (If not, there is no point in checking for FAILED, RUNNING, etc.)

Reply:
The thing is that if an instance failed to come up and thus is in FAILED state, terraform should be able to retry creating it. Running `terraform destroy` to clean up faulty resources and then running `terraform apply` again is the most intuitive way to do this.

That's why the destroy function in the provider should return nil when an instance is in FAILED state.  :) Returning nil tells terraform that the resource has successfully been destroyed and thus it's removed from the .tfstate file.

If trying to destroy a FAILED instance returns an error, the user will have to force remove it from terraform's .tfstate file in order to try again.